File: dotnet/src/Cekat.EventSdk/CekatClient.cs
using System.Net.Http.Headers;
using Cekat.EventSdk.Internal;

namespace Cekat.EventSdk;

/// <summary>
/// Submits events to Cekat. Thread-safe; create one instance per application and reuse it.
/// </summary>
/// <remarks>
/// A client created without an <see cref="HttpClient"/> owns its connection pool and releases it on
/// <see cref="Dispose"/>. An injected <see cref="HttpClient"/> is never disposed by the SDK.
/// </remarks>
public sealed class CekatClient : IDisposable
{
    private static readonly string UserAgent = $"cekat-event-sdk-dotnet/{CekatConstants.Version} dotnet/{Environment.Version}";
    private static readonly MediaTypeHeaderValue JsonContentType = new("application/json");

    private readonly ValidatedOptions _options;
    private readonly HttpClient _http;
    private readonly bool _ownsHttp;
    private readonly IVisitorContext? _visitorContext;
    private readonly Func<double, double> _jitter;
    private readonly Func<TimeSpan, CancellationToken, Task> _delay;
    private readonly TimeProvider _time;

    /// <summary>Creates a client.</summary>
    /// <param name="options">Client configuration; <see cref="CekatClientOptions.AccessToken"/> is required.</param>
    /// <param name="httpClient">An optional caller-owned HTTP client (for proxies, TLS, or handlers). Its own timeout also applies.</param>
    /// <param name="visitorContext">
    /// A request visitor source, such as the ASP.NET Core or Azure Functions context. An explicit
    /// <see cref="EventInput.VisitorId"/> wins, then an <see cref="AsyncLocalVisitorContext"/> scope, then this context.
    /// </param>
    /// <exception cref="CekatValidationException">The options are invalid.</exception>
    public CekatClient(CekatClientOptions options, HttpClient? httpClient = null, IVisitorContext? visitorContext = null)
        : this(options, httpClient, visitorContext, jitter: null, delay: null, time: null)
    {
    }

    internal CekatClient(
        CekatClientOptions options,
        HttpClient? httpClient,
        IVisitorContext? visitorContext,
        Func<double, double>? jitter,
        Func<TimeSpan, CancellationToken, Task>? delay,
        TimeProvider? time)
    {
        _options = OptionsValidator.Validate(options);
        _ownsHttp = httpClient is null;
        _http = httpClient ?? new HttpClient(new SocketsHttpHandler { PooledConnectionLifetime = TimeSpan.FromMinutes(5), AllowAutoRedirect = false })
        {
            Timeout = System.Threading.Timeout.InfiniteTimeSpan,
        };
        _visitorContext = visitorContext;
        _jitter = jitter ?? (upperBound => Random.Shared.NextDouble() * upperBound);
        _delay = delay ?? ((duration, token) => Task.Delay(duration, token));
        _time = time ?? TimeProvider.System;
    }

    /// <summary>Tracks <c>user_registration</c>.</summary>
    /// <exception cref="CekatException">The event was invalid, rejected, or could not be delivered.</exception>
    /// <exception cref="OperationCanceledException"><paramref name="cancellationToken"/> was cancelled.</exception>
    public Task<Acknowledgement> UserRegistrationAsync(EventInput input, CancellationToken cancellationToken = default) =>
        TrackAsync("user_registration", true, () => input, cancellationToken);

    /// <summary>Tracks <c>user_login</c>.</summary>
    /// <exception cref="CekatException">The event was invalid, rejected, or could not be delivered.</exception>
    /// <exception cref="OperationCanceledException"><paramref name="cancellationToken"/> was cancelled.</exception>
    public Task<Acknowledgement> UserLoginAsync(EventInput input, CancellationToken cancellationToken = default) =>
        TrackAsync("user_login", true, () => input, cancellationToken);

    /// <summary>Tracks <c>order_created</c>.</summary>
    /// <exception cref="CekatException">The event was invalid, rejected, or could not be delivered.</exception>
    /// <exception cref="OperationCanceledException"><paramref name="cancellationToken"/> was cancelled.</exception>
    public Task<Acknowledgement> OrderCreatedAsync(EventInput input, CancellationToken cancellationToken = default) =>
        TrackAsync("order_created", true, () => input, cancellationToken);

    /// <summary>Tracks <c>order_paid</c>, sending <paramref name="amount"/> and <paramref name="currency"/> as properties.</summary>
    /// <param name="amount">The paid amount, sent as <c>properties.amount</c>.</param>
    /// <param name="currency">A nonblank currency, sent unchanged as <c>properties.currency</c>.</param>
    /// <param name="input">The event; its properties must not contain <c>amount</c> or <c>currency</c>.</param>
    /// <param name="cancellationToken">Cancels the request or retry delay.</param>
    /// <exception cref="CekatException">The event was invalid, rejected, or could not be delivered.</exception>
    /// <exception cref="OperationCanceledException"><paramref name="cancellationToken"/> was cancelled.</exception>
    public Task<Acknowledgement> OrderPaidAsync(decimal amount, string currency, EventInput input, CancellationToken cancellationToken = default) =>
        TrackAsync("order_paid", true, () => EventPayload.WithOrderPaidProperties(amount, currency, input), cancellationToken);

    /// <summary>Tracks a tenant-defined event.</summary>
    /// <exception cref="CekatException">The event was invalid, rejected, or could not be delivered.</exception>
    /// <exception cref="OperationCanceledException"><paramref name="cancellationToken"/> was cancelled.</exception>
    public Task<Acknowledgement> CustomEventAsync(string eventKey, EventInput input, CancellationToken cancellationToken = default) =>
        TrackAsync(eventKey, false, () => input, cancellationToken);

    /// <summary>Runs <paramref name="action"/> with <paramref name="visitorId"/> as the current visitor.</summary>
    public static async Task<T> WithVisitorIdAsync<T>(string? visitorId, Func<CancellationToken, Task<T>> action, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(action);
        using var scope = AsyncLocalVisitorContext.Push(visitorId);
        return await action(cancellationToken).ConfigureAwait(false);
    }

    /// <summary>Releases the SDK-created connection pool. An injected <see cref="HttpClient"/> is left open.</summary>
    public void Dispose()
    {
        if (_ownsHttp)
        {
            _http.Dispose();
        }
    }

    /// <inheritdoc />
    public override string ToString() => $"CekatClient {{ Endpoint = {_options.Endpoint} }}";

    private async Task<Acknowledgement> TrackAsync(string eventKey, bool isCommon, Func<EventInput> input, CancellationToken cancellationToken)
    {
        // Runs inside the returned task so validation errors fault the task instead of throwing synchronously.
        await Task.Yield();
        var body = EventPayload.Build(eventKey, isCommon, input(), CurrentVisitorId(), _time.GetUtcNow, () => Guid.NewGuid().ToString("D"));
        var attempts = 0;
        while (true)
        {
            cancellationToken.ThrowIfCancellationRequested();
            attempts++;
            ReceivedResponse response;
            try
            {
                response = await AttemptAsync(body, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception error) when (IsTransportFailure(error, cancellationToken))
            {
                if (attempts > _options.RetryCount)
                {
                    throw new CekatTransportException(
                        $"no response from Cekat after {attempts} {(attempts == 1 ? "attempt" : "attempts")} ({error.GetType().Name}); the event may have been received",
                        attempts,
                        error);
                }

                await _delay(RetryPolicy.RetryDelay(attempts, null, _jitter, _time.GetUtcNow()) ?? TimeSpan.Zero, cancellationToken).ConfigureAwait(false);
                continue;
            }

            if (RetryPolicy.IsRetryableStatus(response.StatusCode) && attempts <= _options.RetryCount
                && RetryPolicy.RetryDelay(attempts, response.RetryAfter, _jitter, _time.GetUtcNow()) is { } delay)
            {
                await _delay(delay, cancellationToken).ConfigureAwait(false);
                continue;
            }

            return ResponseDecoder.Decode(response, attempts);
        }
    }

    private string? CurrentVisitorId() =>
        AsyncLocalVisitorContext.Shared.CurrentVisitorId ?? VisitorIdResolver.Normalize(_visitorContext?.CurrentVisitorId);

    private static bool IsTransportFailure(Exception error, CancellationToken cancellationToken) =>
        error is HttpRequestException or IOException || (error is OperationCanceledException && !cancellationToken.IsCancellationRequested);

    private async Task<ReceivedResponse> AttemptAsync(byte[] body, CancellationToken cancellationToken)
    {
        using var attemptCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        attemptCancellation.CancelAfter(_options.Timeout);
        var token = attemptCancellation.Token;

        using var request = new HttpRequestMessage(HttpMethod.Post, _options.Endpoint);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _options.AccessToken);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        request.Headers.TryAddWithoutValidation("User-Agent", UserAgent);
        request.Content = new ByteArrayContent(body);
        request.Content.Headers.ContentType = JsonContentType;

        using var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, token).ConfigureAwait(false);
        var retained = new byte[CekatConstants.MaximumResponseBodyBytes + 1];
        var observed = 0;
        Exception? failure = null;
        try
        {
            await using var stream = await response.Content.ReadAsStreamAsync(token).ConfigureAwait(false);
            int read;
            while (observed < retained.Length && (read = await stream.ReadAsync(retained.AsMemory(observed), token).ConfigureAwait(false)) > 0)
            {
                observed += read;
            }
        }
        catch (Exception error) when (IsTransportFailure(error, cancellationToken))
        {
            failure = error;
        }

        var truncated = observed > CekatConstants.MaximumResponseBodyBytes;
        return new ReceivedResponse(
            (int)response.StatusCode,
            // HTTP/2 and HTTP/3 have no reason phrase; .NET would substitute its own legacy text.
            response.Version.Major < 2 ? response.ReasonPhrase : null,
            response.Headers.NonValidated.TryGetValues("Retry-After", out var retryAfter) ? retryAfter.FirstOrDefault() : null,
            retained.AsSpan(0, Math.Min(observed, CekatConstants.MaximumResponseBodyBytes)).ToArray(),
            truncated,
            observed,
            truncated ? null : failure);
    }
}

File: java/cekat-event-sdk-core/src/main/java/ai/cekat/events/CekatClient.java
package ai.cekat.events;

import ai.cekat.events.error.ApiException;
import ai.cekat.events.error.ResponseDecodeException;
import ai.cekat.events.error.TransportException;
import ai.cekat.events.error.ValidationException;
import ai.cekat.events.internal.PayloadEncoder;
import ai.cekat.events.internal.ResponseDecoder;
import ai.cekat.events.internal.RetryPolicy;
import ai.cekat.events.transport.TransportFailureException;
import ai.cekat.events.transport.TransportRequest;
import ai.cekat.events.transport.TransportResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.OptionalLong;

/**
 * Submits Cekat events synchronously. Construct one client and reuse it; it is thread-safe.
 *
 * <p>Each method returns an {@link Acknowledgement} (accepted for asynchronous processing) or throws
 * {@link ValidationException} before any request, {@link ai.cekat.events.error.AuthenticationException} (401),
 * {@link ai.cekat.events.error.EventDefinitionNotFoundException} (404), {@link ApiException} for other non-200
 * responses, {@link ResponseDecodeException} for an invalid or unreadable 200, or {@link TransportException} when no
 * response was received. Retries may create duplicate events; every retry reuses the call's event ID. Interrupting the
 * calling thread stops the call with {@link InterruptedException}.
 */
public final class CekatClient {
    /** SDK version reported in the User-Agent header. */
    public static final String VERSION = "0.2.0";

    private final String accessToken;
    private final CekatClientOptions options;
    private final PayloadEncoder encoder;

    /**
     * Creates a client with default options.
     *
     * @param accessToken nonblank access token
     */
    public CekatClient(String accessToken) {
        this(accessToken, CekatClientOptions.defaults());
    }

    /**
     * Creates a client.
     *
     * @param accessToken nonblank access token
     * @param options client options
     */
    public CekatClient(String accessToken, CekatClientOptions options) {
        this(accessToken, options, new PayloadEncoder());
    }

    CekatClient(String accessToken, CekatClientOptions options, PayloadEncoder encoder) {
        if (accessToken == null || accessToken.isBlank()) {
            throw new ValidationException("access token must not be blank");
        }
        if (options == null) {
            throw new ValidationException("options must not be null");
        }
        this.accessToken = accessToken;
        this.options = options;
        this.encoder = encoder;
    }

    /**
     * Submits the common user_registration event.
     *
     * @param event event
     * @return acknowledgement
     * @throws InterruptedException when the calling thread is interrupted
     */
    public Acknowledgement userRegistration(Event event) throws InterruptedException {
        return track("user_registration", true, event);
    }

    /**
     * Submits the common user_login event.
     *
     * @param event event
     * @return acknowledgement
     * @throws InterruptedException when the calling thread is interrupted
     */
    public Acknowledgement userLogin(Event event) throws InterruptedException {
        return track("user_login", true, event);
    }

    /**
     * Submits the common order_created event.
     *
     * @param event event
     * @return acknowledgement
     * @throws InterruptedException when the calling thread is interrupted
     */
    public Acknowledgement orderCreated(Event event) throws InterruptedException {
        return track("order_created", true, event);
    }

    /**
     * Submits the common form_submitted event.
     *
     * @param event event
     * @return acknowledgement
     * @throws InterruptedException when the calling thread is interrupted
     */
    public Acknowledgement formSubmitted(Event event) throws InterruptedException {
        return track("form_submitted", true, event);
    }

    /**
     * Submits the common order_paid event. The finite {@code amount} and nonblank {@code currency} are sent as the
     * {@code amount} and {@code currency} properties; the event's properties must not already contain either key.
     *
     * @param amount finite amount, for example a {@link java.math.BigDecimal}
     * @param currency nonblank currency, not validated as a currency code
     * @param event event
     * @return acknowledgement
     * @throws InterruptedException when the calling thread is interrupted
     */
    public Acknowledgement orderPaid(Number amount, String currency, Event event) throws InterruptedException {
        return track("order_paid", true, PayloadEncoder.withOrderPaidProperties(amount, currency, event));
    }

    /**
     * Submits an event with a caller-provided key.
     *
     * @param eventKey nonblank event key
     * @param event event
     * @return acknowledgement
     * @throws InterruptedException when the calling thread is interrupted
     */
    public Acknowledgement customEvent(String eventKey, Event event) throws InterruptedException {
        return track(eventKey, false, event);
    }

    @Override
    public String toString() {
        return "CekatClient[baseUrl=" + options.baseUrl() + ", timeout=" + options.timeout() + ", retryCount=" + options.retryCount() + "]";
    }

    private Acknowledgement track(String eventKey, boolean isCommon, Event event) throws InterruptedException {
        // Encoded once so every retry sends the same event ID and timestamp.
        byte[] body = encoder.encode(eventKey, isCommon, event, VisitorContext.currentVisitorId().orElse(null));
        TransportRequest request = new TransportRequest(options.ingestUri(), Map.of(
                "Authorization", "Bearer " + accessToken,
                "Content-Type", "application/json",
                "User-Agent", "cekat-event-sdk-java/" + VERSION + " java/" + Runtime.version().feature()), body, options.timeout());

        int maximumAttempts = options.retryCount() + 1;
        for (int attempt = 1; ; attempt++) {
            if (Thread.interrupted()) {
                throw new InterruptedException("Cekat event submission interrupted");
            }
            TransportResponse response;
            try {
                response = options.transport().execute(request);
            } catch (TransportFailureException failure) {
                if (attempt >= maximumAttempts) {
                    throw new TransportException("Cekat API request failed before a response was received", attempt, failure);
                }
                backoff(attempt, OptionalLong.empty());
                continue;
            }

            // Retry is decided by status alone, even when the body read failed.
            if (attempt < maximumAttempts && RetryPolicy.isRetryableStatus(response.statusCode())) {
                OptionalLong retryAfter = RetryPolicy.parseRetryAfterMillis(response.header("Retry-After").orElse(null), Instant.now());
                if (retryAfter.isEmpty() || retryAfter.getAsLong() <= RetryPolicy.MAXIMUM_RETRY_AFTER_MILLIS) {
                    backoff(attempt, retryAfter);
                    continue;
                }
                // The server asked for a longer pause than a caller should wait: report it now.
            }
            return ResponseDecoder.decode(response, attempt);
        }
    }

    private void backoff(int retry, OptionalLong retryAfterMillis) throws InterruptedException {
        long jitter = options.jitter().applyAsLong(RetryPolicy.delayBoundMillis(retry));
        long delay = Math.max(jitter, retryAfterMillis.orElse(0));
        options.sleeper().sleep(Duration.ofMillis(delay));
    }
}
